### PR TITLE
Enable and/or searching

### PR DESCRIPTION
**Pull request by Jain-Aditya**
The tool supports searching for multiple hashtags, by entering e.g. HashtagA, HashtagB in the search box. The default behaviour as implemented is for this to return edits which contain either HashtagA or HashtagB. There are times, however, when a user might want to return results that contain HashtagA and HashtagB.

Phabricator Task: T216398

### CODE REVIEW

File: hashtagsv2/hashtags/forms.py [hunk 1]
```diff
@@ -10,6 +12,9 @@ class SearchForm(forms.Form):
 		widget=forms.TextInput(attrs={'placeholder': 'Enter a hashtag',
 			'style': 'width:100%'}))
 
+	type = forms.CharField(label="Type:",
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/forms.py]: This field should be named something else - `type` is an in-built function in Python. Perhaps just something like "search_type" instead.

File: hashtagsv2/hashtags/helpers.py [hunk 2]
```diff
@@ -20,9 +20,19 @@ def hashtag_queryset(request_dict):
 
     hashtag_list = split_hashtags(request_dict['query'])
 
-    queryset = Hashtag.objects.filter(
+    if request_dict['type'] == 'and':
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/helpers.py]: This is causing my local environment to time out when attempting a search. It looks like we're doing potentially thousands of database calls (one for each result returned by the first query).

Rather than looping through rc_ids, try something like the solution at https://stackoverflow.com/questions/769843/how-do-i-use-and-in-a-django-filter/770078#770078. There are better ways, in the existing Django filter behaviour, to do an AND condition.

File: hashtagsv2/hashtags/helpers.py [hunk 3]
```diff
@@ -20,9 +20,19 @@ def hashtag_queryset(request_dict):
...
+            qs = set(Hashtag.objects.filter(rc_id=i).values_list('hashtag',flat=True).distinct())
+            if set(hashtag_list).issubset(qs):
+                selected_rcids.append(i)    
+        queryset = Hashtag.objects.filter(rc_id__in=selected_rcids)
+
+    else:
+        queryset = Hashtag.objects.filter(
         hashtag__in=hashtag_list
-            )
+            )
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/helpers.py]: Looks like you added a bit of random whitespace here :)

File: hashtagsv2/hashtags/templates/hashtags/index.html [hunk 4]
```diff
@@ -29,10 +29,14 @@
                 {{ form.startdate.label }}
                 {{ form.startdate }}
               </div>
-              <div class="three columns">
+              <div class="three columns" style="margin-right: 2em">
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/templates/hashtags/index.html]: This looks good, and in-line, but is probably somewhat confusing to users. We don't even explain that multiple hashtags are supported in the search bar, so seeing an option for and/or isn't likely to make much sense. I'm not sure what the best solution is.

Perhaps we could have an Advanced Search menu that's hidden by default via javascript. Could you try that?

**Review by Samwalton9 — COMMENTED** [review]
See in-line comments :)

**Comment by Jain-Aditya** [comment]
Hi @Samwalton9 I have tried to implement a solution but when I try to search by clicking on Trending tags, it won't work. I guess that is why some of the testcases are not passing. I am not able to figure out the reason what is going wrong. can you please help me out?

File: hashtagsv2/hashtags/helpers.py [hunk 5]
```diff
@@ -20,9 +20,23 @@ def hashtag_queryset(request_dict):
 
     hashtag_list = split_hashtags(request_dict['query'])
 
-    queryset = Hashtag.objects.filter(
-        hashtag__in=hashtag_list
+    if 'search_type' in request_dict:
+        if request_dict['search_type'] == 'and':
+            rcids_for_hashtag=[]
+            for hashtag in hashtag_list:
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/helpers.py:28]: This gets a little hard to follow - could you add some comments explaining the steps involved? :)

File: hashtagsv2/hashtags/forms.py [hunk 6]
```diff
@@ -10,6 +12,9 @@ class SearchForm(forms.Form):
 		widget=forms.TextInput(attrs={'placeholder': 'Enter a hashtag',
 			'style': 'width:100%'}))
 
+	search_type = forms.CharField(label="Type:",
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/forms.py]: This field should have `required=False`, to enable the tool to continue working even when `search_type` isn't in the search request (such as when clicking on a trending tag).

File: hashtagsv2/hashtags/templates/hashtags/index.html [hunk 7]
```diff
@@ -33,6 +33,15 @@
                 {{ form.enddate.label }}
                 {{ form.enddate }}
               </div>
+            </div>
```
Comment (Samwalton9) [inline hashtagsv2/hashtags/templates/hashtags/index.html:36]: Can you also add this to the section which displays when results are listed?

**Review by Samwalton9 — COMMENTED** [review]
This is looking good! I've added a few more in-line comments, including the source of the problem you found where the top hashtag links weren't working.

I'm still not sure how to best explain to users what this and/or field does. Perhaps we could have a small question mark next to the drop-down with some hover text explaining something like "When searching for multiple hashtags, should the results be a subset or union?"

**Comment by Jain-Aditya** [comment]
Thanks @Samwalton9 , I have made the suggested changes.

**Comment by Jain-Aditya** [comment]
Hi @Samwalton9, Just a gentle reminder to review this :)

File: hashtagsv2/hashtags/templates/hashtags/index.html (hunk 4 above)
Comment (prakhargurunani) [reply hashtagsv2/hashtags/templates/hashtags/index.html]:
```suggestion
              <div class="three columns" style="margin-right: 2em">
```
Use percentages rather that a fixed value so users dont face problems in using it .
For centering the ```div```, use this css style -
<div class="three columns" style="transform: translateX(-50%); left: 50%; position: absolute;">

Comment (prakhargurunani) [reply hashtagsv2/hashtags/templates/hashtags/index.html]: Use this css for better view. It automatically collapses in mobile and tablet viewports.
``
<div class="row">
        <div class="col-sm-4">
         
          <div class="col-sm-4">
         
          <div class="col-sm-4">
         
</div>
``

Comment (FirePing32) [reply hashtagsv2/hashtags/templates/hashtags/index.html]: Use this css for better view. It automatically collapses in mobile and tablet viewports.

```
<div class="row">
        <div class="col-sm-4">
         <!-- col 1 here -->
          <div class="col-sm-4">
         <!-- col 2 here -->
          <div class="col-sm-4">
         <!-- col 3 here -->
</div>
```
You will have to use ```bootstrap 3```

Comment (FirePing32) [reply hashtagsv2/hashtags/templates/hashtags/index.html]: ```Perhaps we could have an Advanced Search menu that's hidden by default via javascript. Could you try that?```
Javascript is not needed.
See this -
```
<style>
#stuffToShow {
    display:none;
}

#checkbox {
    opacity:0.01;
}

#checkbox:checked + #stuffToShow {
    display:block;
}
</style>

<label for="checkbox" id="checkboxLabel">Show More</label>
<input type="checkbox" id="checkbox" />
<div id="stuffToShow">
    Happy times
</div>
```

**Comment by Jain-Aditya** [comment]
Resolved merge conflict
